File: PdbReader.cpp
//hello.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <cstdlib>
#include <vector>
#include <sstream>
#include <boost/algorithm/string.hpp>
#include <eigen3/Eigen/Dense>
using namespace std;
using namespace Eigen;
class PdbReader
{
private:
    string filename;
    int model_num;
public:
    vector<string> MODEL;
    PdbReader(string, int);
    int fileread();
    MatrixXd get_matrix(int, int);
};

PdbReader::PdbReader(string s, int num = -1){
    filename = s;
    model_num = num;
}
int PdbReader::fileread(){
    std::ifstream ifs(filename);
    if (!ifs) {
        cerr << "Can't open file" << endl;
        return -1;
    }
    string line;
    vector< vector<string> > all_list;
    vector<std::string> each_list;
    if (model_num == -1){
        while (std::getline(ifs, line)){
            if (line.substr(0, 4) == "ATOM"){
                each_list.push_back(line);
            }
        }
        MODEL = each_list;
    }

    else{
        while (std::getline(ifs, line)){
            if (line.substr(0, 5) == "MODEL"){
                each_list.clear();
            }
            else if(line.substr(0, 4) == "ATOM"){
                each_list.push_back(line);
            }
            else if(line.substr(0, 6) == "ENDMDL"){
                all_list.push_back(each_list);
            }
        }
        MODEL = all_list[model_num - 1];
    }
    return 0;
}

MatrixXd PdbReader::get_matrix(int x_start_col, int element_col){
    int count(-1);
    int N(MODEL.size());
    int weight_sum;
    MatrixXd Q(N, 3);
    MatrixXd M(N, 3);
    for(string line : MODEL){
        count += 1;
        double x, y, z;
        std::vector<std::string> strs;
        // strs is splitted_list of a line by spaces
        boost::split(strs, line, boost::is_space(), boost::algorithm::token_compress_on);
        x = std::stod(strs[x_start_col - 1]);
        y = std::stod(strs[x_start_col]);
        z = std::stod(strs[x_start_col + 1]);
        Vector3d q(x, y, z);
        Q.row(count) = q;
        // calculate moment
        string element(strs[element_col -1]);
        int weight;
        if (element == "H"){
            weight = 1;
        }
        else if (element == "C"){
            weight = 12;
        }
        else if (element == "N"){
            weight = 14;
        }
        else if (element == "O"){
            weight = 16;
        }
        else if (element == "S"){
            weight = 32;
        }
        else{
            cout << "[WARNING] Write the weight of " + element << endl;
        }
        weight_sum += weight;
        RowVector3d moment;
        moment = weight * q;
        M.row(count) = moment;
    }
    RowVector3d c;
    MatrixXd Q_from_center(N, 3);
    c << M.col(0).sum()/weight_sum, M.col(1).sum()/weight_sum,
    M.col(2).sum()/weight_sum;
    Q_from_center = Q.rowwise() - c;
    return Q_from_center;
}

MatrixXd Rotate(MatrixXd Q1, MatrixXd Q2){
    std::vector<int> v{0, 1, 2};
    std::vector<double> D;
    MatrixXd R(3, 3);
    for (auto &i:v){
        for (auto &j:v){
            D.push_back((Q1.array().col(i) * Q2.array().col(j)).sum());
        }
    }
    VectorXd D2 = VectorXd::Map(D.data(), D.size());
    MatrixXd D_matrix = MatrixXd::Map(D2.data(), 3, 3);
    Eigen::JacobiSVD<MatrixXd> svd(D_matrix.transpose(), ComputeThinU | ComputeThinV);
    cout << svd.matrixU() * svd.matrixV().transpose() << endl;
    R = svd.matrixU() * svd.matrixV().transpose();
    return R;
}
// void file_changer(string input_file, int model_num = -1 ,
//                     MatrixXd, string additional_name){
//     std::ifstream ifs(input_file);
//     if (!ifs) {
//         cerr << "Can't open file" << endl;
//         return -1;
//     }
//     string line;
//     vector< vector<string> > all_list;
//     vector<std::string> each_list;
//     int count = -1;
//     if (model_num == -1){
//         while (std::getline(ifs, line)){
//             count += 1
//             if (line.substr(0, 4) == "ATOM"){

//             }
//         }
//         MODEL = each_list;
//     }

//     else{
//         while (std::getline(ifs, line)){
//             if (line.substr(0, 5) == "MODEL"){
//                 each_list.clear();
//             }
//             else if(line.substr(0, 4) == "ATOM"){
//                 each_list.push_back(line);
//             }
//             else if(line.substr(0, 6) == "ENDMDL"){
//                 all_list.push_back(each_list);
//             }
//         }
//         MODEL = all_list[model_num - 1];
//     }
//     return 0;

int main(){
    MatrixXd Q1_from_center;
    PdbReader Reader1("./Wild1_315K_1.pdb");
    Reader1.fileread();
    Q1_from_center = Reader1.get_matrix(6, 11);
    MatrixXd Q2_from_center;
    PdbReader Reader2("./Wild1_315K_2.pdb");
    Reader2.fileread();
    Q2_from_center = Reader2.get_matrix(6, 11);
    MatrixXd R(3, 3);
    R = Rotate(Q1_from_center, Q2_from_center);
    }

//     MatrixXd Q2_from_center;
//     PdbReader Reader2("Wild1_315K_2.pdb", -1);
//     Reader2.fileread();
//     Q2_from_center = Reader2.get_matrix(6, 11);
//     // Eigen::JacobiSVD
// }

File: PdbReader.py
import re
import numpy as np
import os
class PdbReader():

    def __init__(self, pdb_file, model_num=None, x_start_col=None, element_col=None):
        self.pdb_file = pdb_file
        self.model_num = model_num
        self.x_start_col = x_start_col
        self.element_col = element_col
        self.file_name, self.file_ext = os.path.splitext(str(self.pdb_file))

    def get_ATOM(self):
        f = open(self.pdb_file, "r")
        each_model_list = []
        all_model_list = []
        if self.model_num == None:
            for line in f:
                if line.startswith("ATOM"):
                    each_model_list.append(str(line.strip()))
            f.close()
            MODEL = each_model_list
        else:
            for line in f:
                if line.startswith("MODEL"):
                    each_model_list = []

                elif line.startswith("ATOM"):
                    each_model_list.append(line)

                elif line.startswith("ENDMDL"):
                    all_model_list.append(each_model_list)
            MODEL = all_model_list[int(self.model_num) - 1]
            f.close()
        self.MODEL = MODEL
        return(MODEL, all_model_list)

    def move_center(self, MODEL):
        q_array = np.array([])
        weight_q_array = np.array([], dtype = 'float64')
        moment_array = np.array([], dtype = 'float64')
        Total_weight = 0
        for line in MODEL:
            if self.x_start_col == None:
                x = line.split()[6]
                y = line.split()[7]
                z = line.split()[8]
            else:
                x = line.split()[self.x_start_col]
                y = line.split()[self.x_start_col+1]
                z = line.split()[self.x_start_col+2]
            if self.element_col == None:
                element= line.split()[11]
            else:
                element= line.split()[self.element_col]

            q = np.array([x, y, z], dtype = 'float64')
            q_array = np.append(q_array, q)

            if element == 'H':
                weight = 1
            elif element == 'C':
                weight = 12
            elif element == 'N':
                weight = 14
            elif element == 'O':
                weight = 16
            elif element == "S":
                weight = 32
            else:
                print("[[[WARNING]]] Please write the weight of Element:" + element)
            moment = q * weight
            moment_array = np.append(moment_array, moment)
            Total_weight += weight
            weight_q = q * (weight**(1/2))
            weight_q_array = np.append(weight_q_array, weight_q)
        q_matrix = q_array.reshape(int(q_array.shape[0]/3), 3)
        weight_q_matrix=  weight_q_array.reshape(int(weight_q_array.shape[0]/3), 3)
        moment_matrix = moment_array.reshape(int(moment_array.shape[0]/3), 3)
        center = np.array([sum(moment_matrix[:, 0]), sum(moment_matrix[:, 1]), sum(moment_matrix[:, 2])]) / Total_weight
        q_matrix_from_center = q_matrix - center
        self.q_matrix_from_center = q_matrix_from_center
        return(q_matrix_from_center)

    def file_changer(self):
        count = -1
        f = open(self.pdb_file, "r")
        new_file = open(self.file_name + '_from_center' + self.file_ext, 'w')
        if self.model_num == None:
            for line in f:
                if line.startswith("ATOM"):
                    count += 1
                    if self.x_start_col == None:
                        x = line.split()[6]
                        y = line.split()[7]
                        z = line.split()[8]
                    else:
                        x = line.split()[self.x_start_col]
                        y = line.split()[self.x_start_col+1]
                        z = line.split()[self.x_start_col+2]
                    n = 0
                    new_line = line
                    for be in str(x), str(y), str(z):
                        len_be = len(be)
                        af = str(round(self.q_matrix_from_center[count][n], 3))
                        len_af = len(af)
                        if len_be == len_af:
                            new_line = new_line.replace(be, af)
                        elif len_be > len_af:
                            new_line = new_line.replace(be, ' '*(len_be-len_af)+af)
                        elif len_be < len_af:
                            new_line = new_line.replace(' '*(len_af-len_be) + be, af)
                        n += 1
                else:
                    new_line = line
                print(new_line.strip())
                new_file.write(new_line)
        f.close()
        new_file.close()

# Rotate PdbClass1 matrix to Pdbclass2 and output the rotated matrix to an file.
def Rotate(PdbReaderClass1, PdbReaderClass2):
    D = []
    for i in [0, 1, 2]:
        for j in [0, 1,2 ]:
            D.append(np.sum(PdbReaderClass1.q_matrix_from_center[:, i] * PdbReaderClass2.q_matrix_from_center[:, j]))
    D = np.array(D)
    D_matrix = D.reshape(3, 3)
    U, S, V = np.linalg.svd(D_matrix, full_matrices=False)
    R = np.dot(U, V)
    Rotated = np.dot(PdbReaderClass1.q_matrix_from_center, R)
    PdbReaderClass1.file_changer(Rotated, "_rotated")